bin/replace-secrets.js

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Replaces redacted secret environment variables in an app
+ * directory's `redacted.env` and outputs `.env` with the decrypted values of
+ * the secrets.
+ */
+
+const fs = require('fs');
+const path = require('path');
+const YAML = require('yaml');
+const md5 = require('md5');
+
+const OUTPUT_ENV_FILE = '.env';
+const REDACTED_ENV_FILE = 'redacted.env';
+const CLOUD_BUILD_FILE = 'cloud_build.yaml';
+
+/**
+ * Reads the Cloud Build config file to identify secrets in the environment.
+ *
+ * @param {string} appDir path to app directory.
+ * @return {Array<string>} list of environment variable names.
+ */
+function identifySecrets(appDir) {
+  const yamlFile = path.join(appDir, CLOUD_BUILD_FILE);
+  try {
+    const yamlContents = fs.readFileSync(yamlFile).toString('utf8');
+    const config = YAML.parse(yamlContents);
+    const secrets = Object.keys(config.secrets[0].secretEnv);
+    console.log(`Identified the following secret env keys: ${secrets}`);
+    return secrets;
+  } catch (e) {
+    // Handle the case where parsing secrets from the cloud build file fails.
+    console.warn('No secret keys were found in cloud_build.yaml.');
+  }
+
+  return [];
+}
+
+/**
+ * Reads the redacted env file and creates a `.env` file including secrets.
+ *
+ * @param {string} appDir path to app directory.
+ */
+function replaceSecrets(appDir) {
+  const envFile = path.join(appDir, REDACTED_ENV_FILE);
+  const envFileContents = fs.readFileSync(envFile).toString('utf8');
+
+  const replacedContents = envFileContents
+    .split('\n')
+    .map(line => {
+      for (const secret of identifySecrets(appDir)) {
+        if (line.startsWith(`${secret}=`)) {
+          const secretVal = process.env[secret];
+          // The below use of MD5 is safe because it's not actually used to
+          // encrypt the secret; only a substring of the MD5 is displayed as a
+          // fingerprint for devs to verify that the build environment has the
+          // correct values provided for secrets.
+          const partialHash = md5(secretVal).substr(0, 6); // lgtm [js/weak-cryptographic-algorithm]
```

Comment:
Why not just use SHA to stop lgtm from complaining?

Reply:
Sure, why not. Donezo!